### ISSUE TITLE
Custom injector for inherited page properties

### ISSUE THREAD

**Issue by toniedzwiedz** [Adobe-Consulting-Services/acs-aem-commons#660]
On one of my projects, I've been steadily rewriting long and obscure adapter factories in favour of Sling Models. The library offers a useful set of capabilities further extended by the AEM Commons Injector. I can often get away with very clean, declarative code based on annotations, with almost no manual, boilerplate code.

One case, in which I need to put some logic in my `@PostConstruct` method is when reading inherited page properties. I usually `@Inject` an instance of `Resource` and adapt it to an [`InheritanceValueMap`](https://docs.adobe.com/docs/en/aem/6-0/develop/ref/javadoc/com/day/cq/commons/inherit/InheritanceValueMap.html) and then read the properties and assign them to the right fields. This works and doesn't take very long but feels somewhat repetitive.

```
@Model(adaptables=Resource.class)
public class MyModel {

    @Inject
    private Resource resource;

    private String inheritedProperty;

    @PostConstruct
    public void postConstruct() {
        InheritanceValueMap ivm = new HierarchyNodeInheritanceValueMap(resource);
        this.inheritedProperty = ivm.getInherited("inheritedProperty", String.class);
    }

    //getters
}
```

How about an additional binding annotation to mark the fields to be obtained through an `InheritanceValueMap` and a custom injector to make this easier?

Maybe something along the lines of:

```
@Model(adaptables=Resource.class)
public class MyModel {

    @Inject
    @InheritedProperty
    private String inheritedProperty;

    //getters
}
```

Or maybe using the `@Source` annotation to specify the injector, if some performance could be gained that way...

```
@Model(adaptables=Resource.class)
public class MyModel {

    @Inject
    @Source("inherited-properties")
    private String inheritedProperty;

    //getters
}
```

Or maybe there's already a way to do that?

Your thoughts?

**Comment by schoudry** [comment]
this doesnt work?

    ```
@Inject @Via("resource") @Named("dioStyle") @Optional
    public String catalogPath;
```

**Comment by toniedzwiedz** [comment]
@schoudry never occurred to me that this combination would get me an inherited property. I'll try it tomorrow.

I must say it looks a bit cryptic, can you explain how it reads an inherited property?

**Comment by schoudry** [comment]
ignore, never tried "InheritanceValueMap", would be a good to have such injector if the above code doesnt work - certainly works for valuemap

**Comment by toniedzwiedz** [comment]
@schoudry ah, I see. I know it works for a value map OOTB, it's the inheritance I'd like to keep sweet and concise.

**Comment by cpilsworth** [comment]
@toniedzwiedz I like this idea, seems like a candidate for Sling rather than this project though - there is nothing AEM specific about it.

**Comment by justinedelson** [comment]
@cpilsworth the InheritanceValueMap class is AEM-specific (unfortunately IMHO).

**Comment by toniedzwiedz** [comment]
@cpilsworth what Justin just said. The `InheritanceValueMap` interface and its implementations belong to  the `com.day.cq.commons.inherit` package.

Moreover, the `HierarchyNodeInheritanceValueMap` is based on the concept of pages that, AFAIK, is also AEM specific.

These are the only reasons I posted the suggestion here rather than in the Apache Sling Jira

This particular implementation has been the most prevalent in the projects I worked on. The interface used primarily with page properties (although it's often a good idea to separate inheritance-driven configurations from the main content structure for more flexible replication and caching options) and rarely used in different contexts. I'd love to hear about your observations in this regard.
